zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,927 @@
+use core::fmt::Debug;
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::fmt;
+
+use base64;
+
+use nom::{
+    character::complete::char, combinator::complete, multi::separated_list, sequence::preceded,
+};
+
+use zcash_primitives::{consensus, legacy, primitives, transaction::components::Amount};
+
+use crate::encoding::{
+    decode_payment_address, decode_transparent_address, encode_payment_address,
+    encode_transparent_address,
+};
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum Address {
+    TransparentAddress(legacy::TransparentAddress),
+    SaplingAddress(primitives::PaymentAddress),
+}
+
+#[derive(Debug)]
+pub enum AddrParseError {
+    TAddrError(bs58::decode::Error),
+    SAddrError(bech32::Error),
+}
+
+impl Address {
+    pub fn render<P: consensus::Parameters>(&self, params: &P) -> String {
+        match self {
+            Address::TransparentAddress(t) => encode_transparent_address(
+                &params.b58_pubkey_address_prefix(),
+                &params.b58_script_address_prefix(),
+                &t,
+            ),
+            Address::SaplingAddress(s) => {
+                encode_payment_address(&params.hrp_sapling_payment_address(), &s)
+            }
+        }
+    }
+
+    /// Parser for shielded (Sapling) and transparent ZCash addresses.
+    pub fn parse<P: consensus::Parameters>(
+        params: &P,
+        addr_str: &str,
+    ) -> Result<Option<Address>, AddrParseError> {
+        if addr_str == "" {
+            Ok(None)
+        } else {
+            let t_res = decode_transparent_address(
+                &params.b58_pubkey_address_prefix(),
+                &params.b58_script_address_prefix(),
+                addr_str,
+            )
+            .map(|t| t.map(Address::TransparentAddress))
+            .map_err(AddrParseError::TAddrError);
+
```

Comment:
Could you short circuit with `return` here such that `s_res` is not always computed?

Is there a security guarantee that is enforced by running this `fn` in constant time?

Reply:
This has actually *just* been subsumed by moving the existing parsing in the zcash_client_sqlite crate up the hierarchy, so this has now been deleted. :)